talk/morelanguage/morestl.tex

```diff
@@ -25,6 +25,26 @@ \subsection{More STL}
   \end{block}
 \end{frame}
 
+\begin{frame}[fragile]
+  \frametitlecpp[17]{std::optional}
+  \scriptsize
+  \begin{cppcode}
+    std::optional<Phone> parse\_phone(std::string\_view in)
+    {
+      if (is\_valid\_phone(in)) {
+        return std::optional<Phone>(in);
+      }
+      return std::nullopt; // {} also default constructs to nullopt
+    }
+
+    ...
+    if (auto v = parse\_phone(get\_input());
+        v.is\_valid()) {
+        process\_phone(v.value());
+    }
```

Comment:
`std::optional` is contextually convertible to bool, so this should suffice:
```suggestion
    if (auto v = parse\_phone(get\_input())) {
        process\_phone(v.value());
    }
```
Nevertheless, it's good to point out that `has_value() -> bool` also exists.

Reply:
Agreed that the contextual conversion to `bool` should be demonstrated. But again I would avoid mixing it with other features that may not yet been assimilated, here the usage of init-statement in particular and the fact that `operator=` returns its result. So let's maybe simplify to 
```cpp
  auto v = parse_phone(get_input());
  if (v) {
     process_phone(v.value());  // or *v, equivalent to v.value()
  }
```

Reply:
You can do as you like. I still prefer to have
```c++
  if (auto v = parse_phone(get_input())) {
     process_phone(v.value());  // or *v, equivalent to v.value()
  }
```
on the slide somewhere, because this is what I would like the students to remember and write in their code bases.

Reply:
sure makes sense

Reply:
So my only comment on this is since at first glance this is an assignment instead of an expression, (but if RHS is null the if condition would fail), would this be intuitive to students?